Support for incremental batch updates and sample insertion

jtnystrom opened the pull request:
## Summary

Since just before this PR, we have had support for per-batch attributes, which are now enabled by default when data is uploaded from the Scalatra servlet. We have also had, since a long time ago, support for incremental data insertion into batches. However, there was no way of triggering the incremental insertion using the HTTP API (only command-line tools could do it). This PR adds the ability to incrementally insert new samples into existing batches ("append"), as well as partially update metadata in batches (i.e. for some samples only rather than the whole batch). These operations work with both per-batch attributes and with old-style systemwide attributes.
The intention is to allow data maintainers to gradually build up a dataset by adding a few samples at a time, or by redefining old samples with additional or changed attributes.

## Details

The affected APIs are POST /batch and PUT /batch. This change adds to the capabilities of these requests without removing any old functionality.
POST/batch previously allowed new batch upload only. This PR adds the ?append flag which allows the request to add new samples to an existing batch. In this case, an existing batch ID should be specified, and samples not referenced by the uploaded data and metadata files will be left intact in the batch. The additional parameters (dataset, enabled instances, comment etc) for this call should also be specified and will also be updated in the database for the existing batch.
PUT /batch previously could update the metadata of an entire batch only. This change allows it to also function when the metadata is partial, i.e. for some subset of the samples in the batch. No special parameter is needed as this is just broadening the scope of the old operation.

Some technical details:
* Samples are added, removed or updated on a row-major basis in the metadata (whole samples at a time, rather than e.g. column-major: whole attributes at a time)
* Appending samples or updating metadata, even when the update is partial, can add or redefine custom attributes, but not delete them. Thus, an increasing number of attributes can be built up over time. However, attributes that are tracked but that are not in use by any sample will remain hidden from users. They can be completely deleted by removing and reinserting the batch, if desired.
* The concept of "System" attributes that cannot be redefined or removed (e.g. sample_id, treatment) was introduced.

## Validation

A file OTGTool/testData/meta_customAttributes.tsv was created to test adding and renaming custom attributes in a minimal metadata file (accompanied by data.csv in the same folder)
Adding and renaming attributes for a subset of samples was tested by editing this file and creating various versions of it, for subsets of the samples, then updating the batch. New attributes and attribute names showed up in the Angular GUI as expected.
In the same folder, data.csv and meta.tsv were split into two parts each containing a subset of the samples of the full batch (Tunicamycin 48 samples), and it was verified that the two parts could be inserted separately with ?append=true (for the second part) to form the full batch, and that this data was correctly inserted.

jtnystrom committed:
Allow partial update of batch metadata (for some samples only)

jtnystrom committed:
Delete batch timestamp when metadata is partially updated (so that we don't end up with multiple timestamps)

jtnystrom committed:
When querying per-batch attributes, only return those that have defined values for some sample

jtnystrom committed:
Rename some methods for clarity

jtnystrom committed:
Introduce the notion of "system attributes" in AttributeSet (which may not be redefined) and add redefineOrCreate() method for attributes

jtnystrom committed:
When updating batch metadata, redefine those custom attributes that are included in the metadata.
This allows attribute title and type to be changed.
For readability, remove AttributeSetHelper.

jtnystrom committed:
Bugfix for previous commit

jtnystrom committed:
Add "append" parameter to post /batch for appending samples to existing batches

jtnystrom committed:
Simplify and comment some methods related to updateMetadata in BatchManager, make sure append/update cases are handled correctly

jtnystrom committed:
Halt(400) in post /batch if data files are not specified, rather than just throw an exception

jtnystrom committed:
Simplify appending/updating metadata. Appending to a batch is now conceptually the same as updating it, except that samples may be added.
Check that the samples specified in the expression data and the metadata when inserting samples are precisely the same, and refuse to process updates otherwise.

jtnystrom committed:
Add comment

jtnystrom commented:
These features were tested in the following way:

* A file OTGTool/testData/meta_customAttributes.tsv was created to test adding and renaming custom attributes in a minimal metadata file (to be accompanied by data.csv in the same folder)
* Adding and renaming attributes for a subset of samples was tested. New attributes and attribute names showed up in the Angular GUI as expected.
* In the same folder, data.csv and meta.tsv were split into two parts each containing a subset of the samples of the full batch (Tunicamycin 48 samples), and it was verified that the two parts could be inserted separately with ?append=true to form the full batch.

jtnystrom committed:
Add comments

jtnystrom committed:
Correct comment

yujiberra commented on OTGTool/src/main/scala/t/db/file/TSVMetadata.scala:
```diff
@@ -57,12 +56,17 @@ object TSVMetadata {
+              } else if (!attrSet.hasId(normalizedID)) {
+                warningHandler(s"attribute $normalizedID not found, creating new as type $attribType")
+              } else {
+                warningHandler(s"Attribute $normalizedID exists and may not be redefined")
```
It might make more sense for this to be an exception?

jtnystrom replied:
Yes, but then I would have to create additional logic to really identify the illegal circumstances, since specifying e.g. sample_id and platform is part of normal data insertion. For now, I will change the message to sound less threatening instead since this is not really something wrong that the user is doing.

yujiberra replied:
Right, yes, I think I misunderstood what those cases meant, thanks for the change!

I think my confusion partly stems from it being called a `warningHandler`

jtnystrom replied:
It's indeed misleading. Let's fix that too.

yujiberra commented on OTGTool/src/main/scala/t/manager/BatchManager.scala:
```diff
@@ -568,10 +586,27 @@ class BatchManager(context: Context) {
+   *                 If given, only samples in the metadata, as well as all custom attributes in the batch will
+   *                 be deleted.
+   */
+  def deleteBatchRDF(title: String, metadata: Option[Metadata]) =
```
Might make more sense to take an array of samples rather than a metadata?

yujiberra commented on Toxygates/src/panomicon/UploadHandling.scala:
```diff
@@ -55,15 +55,18 @@ class UploadHandling(context: Context) {
 
     val existingBatches = new BatchStore(context.config.triplestoreConfig).getList()
-    if (existingBatches.contains(batch.id) && !mayAppendBatch) {
+    if (existingBatches.contains(batch.id) && !append) {
       throw BatchUploadException.badID(
         s"The batch $batch already exists and appending is not allowed. " +
           "Please choose a different name.")
+    } else if (!existingBatches.contains(batch.id) && append) {
+      throw BatchUploadException.badID(
+        s"The batch $batch does not exist, but appending was requested.")
```
:+1:

yujiberra commented on OTGTool/src/main/scala/t/manager/BatchManager.scala:
```diff
@@ -568,10 +586,27 @@ class BatchManager(context: Context) {
+          bs.deleteCustomAttributes(title)
+          //Also delete the old timestamp, in anticipation of a new one
+          bs.deleteTimestamp(title)
+          //Delete all data associated with the batch
+        case _ => bs.delete(title)
```
I appreciate the comments in this method in general!
I think the indentation on this comment is off, though, if it's meant to apply to the next line

yujiberra reviewed (commented):
Thanks for your work! Sorry it took me so long to review.

It feels a bit difficult to review PRs to a codebase that's in this state because there's so much that could be improved, but overall I think it looks pretty good.

> POST/batch 
> The additional parameters (dataset, enabled instances, comment etc) for this call should also be specified and will also be updated in the database for the existing batch.

As a REST API it would probably be better if these arguments were optional, and not specifying them would preserve the relevant batch properties. But of course those properties would be required for a new batch. So actually overall I think it would make the most sense for batch creation and batch appending to be separate endpoints.

I think that's the one major change I'd want to suggest, or at least get your thoughts on. Other than that I left a few minor comments.

Also, just wanted to clarify the below:

> - Samples are added, removed or updated on a row-major basis in the metadata (whole samples at a time, rather than e.g. column-major: whole attributes at a time)

I'm not sure what this means, especially whether it's relevant to API output or if it's more referring to the implementation. So, imagine I append to an existing batch a metadata file that has just one sample, and (besides system attributes) only includes one metadata column (which is new) and has a non-empty value for the sample. 

There are then two separate sets of questions here:

- What happens if I get the metadata for this batch from the API
  - For the new sample, what values do I see for existing attributes, that were not specified in the new metadata file
  - For existing samples, what values do I see for the new attribute, for which the older samples don't have values
- Analogous questions to the above for what is stored in the batch graph

yujiberra commented:
Actually a quick though about this:

> So actually overall I think it would make the most sense for batch creation and batch appending to be separate endpoints.

On further reflection I think what would make the most sense would be two endpoints, 

- one (POST) to create new batches
- one (PUT) to modify existing batches (so both 'batch update' and 'batch appending' would be here)

assuming I'm not missing something that makes this infeasible

jtnystrom committed:
BatchManager cleanups

jtnystrom committed:
When new metadata refers to existing non-redefinable attributes, emit a less severe message since this is normal

jtnystrom committed:
Rename warningHandler to logMessage in TSVMetadata, since this captures the intention better

jtnystrom commented:
> 
> There are then two separate sets of questions here:
> 
>     * What happens if I get the metadata for this batch from the API
>       
>       * For the new sample, what values do I see for existing attributes, that were not specified in the new metadata file
>       * For existing samples, what values do I see for the new attribute, for which the older samples don't have values
> 
>     * Analogous questions to the above for what is stored in the batch graph

In general, the API will not return any data for attributes that are not specified. So in case 1, the new sample, you will not see any values for those attributes, even if you request them. They will not be in the returned data, not even as empty strings or nulls etc.
The situation is the same for case 2, the existing values and the new attributes. This reflects what is stored in the batch graph in these cases (i.e. nothing for undefined attributes for a given sample). 

However, for a truly "empty" attribute, the batch graph will maintain a (possibly useless) record that the attribute itself exists, even when no samples have defined values for it (however, there should be no way to observe an empty attribute from the API, so this is a minor technical point). It will remain in place in the graph until the entire batch is deleted or until the attribute is redefined.

Having said this, the above is my intention with this design, and it works like this today for the purposes of the Angular application. However, it may take a little while to roll this behavior out to every single API request and to iron out some wrinkles.

jtnystrom commented:
> I'm not sure what this means, especially whether it's relevant to API output or if it's more referring to the implementation.



> > * Samples are added, removed or updated on a row-major basis in the metadata (whole samples at a time, rather than e.g. column-major: whole attributes at a time)
> 
> I'm not sure what this means, especially whether it's relevant to API output or if it's more referring to the implementation.

I think the point was poorly phrased, let me try to express what this may mean to API users in practice:
When you describe a sample in metadata that you add (or update) you are expected to be describing all the known attributes for that sample. So you are always redefining a sample "in full" when you refer to it, and this is the smallest unit of data updates.
By contrast there is no way to just redefine all values for an attribute in full (conceptually speaking). This can only be done by updating it in all samples that make use of it.

jtnystrom commented:
> > * Samples are added, removed or updated on a row-major basis in the metadata (whole samples at a time, rather than e.g. column-major: whole attributes at a time)
> 
> I'm not sure what this means, especially whether it's relevant to API output or if it's more referring to the implementation.

I think the point was poorly phrased, let me try to express what this may mean to API users in practice:
When you describe a sample in metadata that you add (or update) you are expected to be including all the known attributes for that sample (that you want to preserve). So you are always redefining a sample "in full" - a full row - when you refer to it, and this is the smallest unit of data updates.
By contrast there is no way to just redefine all values for an attribute in full - a full column - (conceptually speaking). This can only be done by updating it in all samples that make use of it.

jtnystrom committed:
Rather than have an append flag on POST /batch, let PUT /batch handle all batch updates including appending samples.
Sample appending is now done by passing optional expression data files to PUT /batch.
Refactor BatchManager somewhat, introducing the inner class BatchDataOps, to avoid having lots of methods with very long argument lists.

jtnystrom committed:
When inserting time/dose series data, check if all necessary attributes are present in metadata, and if not skip this step

jtnystrom commented:
With these latest commits, the sample append operation has been moved to PUT /batch like you said, which I think makes more sense. I was also able to clean up a few other things.

yujiberra commented on Toxygates/src/panomicon/ScalatraJSONServlet.scala:
```diff
@@ -415,14 +412,19 @@ class ScalatraJSONServlet(scontext: ServletContext) extends ScalatraServlet
+   * or partial (for just some samples). In the latter case, existing samples that are not referenced in the new
+   * metadata will be left unchanged.
+   *
+   * If exprData is specified, new samples are appended to the batch. Metadata must then also be specified, and
+   * must describe exactly the same samples that the metadata describes.
```
This behavior sounds strange to me; does that mean there isn't a way to update expression data for existing samples?

jtnystrom replied:
No, this is possible: the calls that allow "appending" also allow overwriting existing sample data (so "append" is really the wrong word). I will try to improve the comments and method naming.

jtnystrom replied:
This has now been done, I tried to use a different word than 'append' to indicate that updating (overwriting) expression data is also possible. I also tested it.

yujiberra replied:
Right, so I guess that's [0c7d591](https://github.com/Toxygates/Panomicon/pull/1/commits/0c7d591a0e14064aded015fc69a44ccd9c29bf65), makes sense. Thanks!

yujiberra commented on OTGTool/src/main/scala/t/manager/BatchManager.scala:
```diff
@@ -235,75 +239,124 @@ class BatchManager(context: Context) {
+              conversion: Option[ExpressionConverter] = None) =
+    new BatchDataOps(batch, metadataFile, customAttributes, conversion)
+
+  class BatchDataOps(batch: Batch, metadataFile: String, customAttributes: Boolean,
+                     conversion: Option[ExpressionConverter]) {
```
I'm not sure how I feel about this architecture, especially if it exists solely to make some argument lists shorter. If that's the main goal, then I think making something like

`case class BatchOperationContext(batch: Batch, customAttributes: Boolean, conversion: Option[ExpressionConverter)`

and putting that in various methods' argument lists would be just as effective and a bit easier to understand. 

(The reason I left out `metadata` is because I don't think it conceptually belongs with the other elements. Making a case class with metadata, expression, and (optional) calls could make sense too. Also I only have a vague idea of what `conversion` is so I'm not _that_ confident about it belonging.)

Hmm yeah it's possible that making a class like `BatchDataOps` is actually a common architectural pattern that I'm not familiar with, and that is part of the reason for my reaction, but I do feel that whatever it is that it encapsulates in this case is not that conceptually coherent.

jtnystrom replied:
I think I agree with you, conceptually this is a little messy. I will replace this with a different architecture based on ExprDataInput and MetadataInput, as we discussed previously.

jtnystrom replied:
This has now been changed.

yujiberra replied:
Right, so I guess that's [0c7d591](https://github.com/Toxygates/Panomicon/pull/1/commits/0c7d591a0e14064aded015fc69a44ccd9c29bf65), makes sense. Thanks!

yujiberra replied:
I guess this is done in [b2e610b](https://github.com/Toxygates/Panomicon/pull/1/commits/b2e610b746f9c79d5eaa6b207d396df4a7e574ad) along with some extra refactoring, looks good overall (though I think the commit could have been split up a little)

jtnystrom replied:
Correct, thanks

yujiberra commented on Toxygates/src/panomicon/ScalatraJSONServlet.scala:
```diff
@@ -431,12 +433,18 @@ class ScalatraJSONServlet(scontext: ServletContext) extends ScalatraServlet
     val comment = params.get("comment").getOrElse("")
     val publicComment = params.get("publicComment").getOrElse("")
```
Can you check that updating the publicComment here works? I started trying to test this PR, but I am unable to update the public comment. I put a println to make sure that I was correctly passing the parameter so it shouldn't be a frontend issue. Also, I don't think this should matter, but this was a batch update without a metadata file.

jtnystrom replied:
This was indeed a bug, and the latest commits should fix that.

yujiberra replied:
It seems like both [3a48469](https://github.com/Toxygates/Panomicon/pull/1/commits/3a48469797be1dcd1e2dd8fc8fadd915f2d203ba) and [01982f3](https://github.com/Toxygates/Panomicon/pull/1/commits/01982f38cb844c9b027ebf71d251931763dba40b) have to do with this, but there's also a little more going on there?

jtnystrom replied:
In hindsight, I think they should have been structured differently.
I committed 3a48649 after identifying a bug that I needed to fix (t:comment wasn't being updated). Then I ended up overhauling the same functionality even more in 01982f3, and I found a way to simplify what was being done by using updateSinglePredicate (which removes the single existing value for a predicate before adding a replacement value). This removed the need for the method "deleteBatchProperties".

yujiberra reviewed (commented):
@jtnystrom 
Thanks for the changes! Apologies for the delay in reviewing again. Most of the updates look good, but I have a couple questions and also something that came up as I started to do testing.

Thanks also for your explanations, I just wanted to respond to this:

> When you describe a sample in metadata that you add (or update) you are expected to be including all the known attributes for that sample (that you want to preserve). So you are always redefining a sample "in full" - a full row - when you refer to it, and this is the smallest unit of data updates.

I think it would be nice to be able to update e.g. just one attribute for a set of samples, without having to reupload full metadata for those samples. Of course, if we changed the behavior of the PUT endpoint to accommodate that, there would no longer be any way to delete an attribute value for one or more samples. I guess there could either be a toggle between two modes of operations for the PUT operation, or there could be another endpoint for partial metadata deletion.

That's probably a moderate-sized undertaking though so I don't think it needs to be done as part of this PR.

jtnystrom committed:
Refactoring BatchManager for simpler add/update API - work in progress

jtnystrom committed:
Improve comments and method naming for clarity

jtnystrom committed:
Update both batch timestamp and comment when batch metadata is updated (BatchManager.deleteBatchRDF)

jtnystrom committed:
Simplify how batch properties are updated.
Correctly set public comment when these change.

jtnystrom commented on OTGTool/src/main/scala/t/manager/BatchManager.scala:
```diff
@@ -662,8 +665,6 @@ class BatchManager(context: Context) {
           //If we are updating metadata, then these will be reinserted after the attribute set has been updated.
           bs.deleteCustomAttributes(title)
-          //Delete batch properties that will be replaced
-          bs.deleteBatchProperties(title)
```
This call is no longer needed because updateSinglePredicate will delete the old value as it inserts the new one

jtnystrom commented on OTGTool/src/main/scala/t/sparql/ListManager.scala:
```diff
@@ -161,19 +161,31 @@ abstract class ListManager[T](config: TriplestoreConfig) extends Closeable {
 
   def setComment(name: String, comment: String): Unit = {
-    updateSinglePredicate(name, "t:comment", "\"" + comment + "\"")
+    if (comment.contains('"')) {
```
This contains(") checking should have been here all along, so I think it was an opening for bugs or weird data issues. I decided to fix it while I was working here.

yujiberra reviewed (approved):
Looks good, thanks for the changes!

I ended up discussing the changes directly with Johan, but just to summarize, the last batch of changes addressed all of my concerns, and to make reviewing easier I suggested doing the following in future PRs:
- when changes are made in response to a review comment, to insert the relevant commit hash or link in the response
- breaking up refactor commits like [01982f3](https://github.com/Toxygates/Panomicon/pull/1/commits/01982f38cb844c9b027ebf71d251931763dba40b) a bit into conceptually cohesive individual commits where possible
